### PR TITLE
fix(plugin-meetings): refactor addMedia and add separate timeout for sdp exchange

### PR DESCRIPTION
**Pull request by jor-row**
# COMPLETES [SPARK-480556](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-480556)

## This pull request addresses

Refactor of addMedia to change it to an async function and move some of the contents into private methods in preparation for https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-480552

Add timeout for sdp exhange which I have set to 10 seconds

Change timeout PC_BAIL_TIMEOUT from 15 seconds to 10 seconds

Turn off mercury OFFLINE listener in unsetPeerConnections()

### Change Type

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [x] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [ ] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

File: docs/samples/webex.min.js
[the remark this answers is not in the document]
Comment (jor-row) [reply docs/samples/webex.min.js:1]: Done. Why are changes to these files allowed to be committed when they are explicitly in the gitignore?

File: packages/@webex/plugin-meetings/src/constants.ts [hunk 1]
```diff
@@ -190,7 +190,8 @@ export const ICE_FAIL_TIMEOUT = 3000;
 
 export const RETRY_TIMEOUT = 3000;
 
-export const PC_BAIL_TIMEOUT = 15000;
+export const PC_BAIL_TIMEOUT = 10000;
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/constants.ts:193]: Renaming it to "ICE_AND_DTLS_CONNECTION_TIMEOUT". It's a little verbose but I'd rather it be verbose than unclear.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 2]
```diff
@@ -5589,28 +5642,285 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+      await mc.initiateOffer();
+
+      await this.waitForRemoteSDPAnswer();
+
+      this.handleMediaLogging(this.mediaProperties);
+      LoggerProxy.logger.info(`${LOG_HEADER} media connection created`);
+    } catch (error) {
+      LoggerProxy.logger.error(
+        `${LOG_HEADER} Error adding media , setting up peerconnection, `,
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:5877]: I agree, changing it to your suggestion.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 3]
```diff
@@ -5589,28 +5642,285 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * @private
+   * @param {RemoteMediaManagerConfiguration} [remoteMediaManagerConfig]
+   * @param {BundlePolicy} [bundlePolicy]
+   * @returns {Promise<void>}
+   */
+  private async establishMediaConnection(
+    remoteMediaManagerConfig?: RemoteMediaManagerConfiguration,
+    bundlePolicy?: BundlePolicy
+  ): Promise<void> {
+    const LOG_HEADER = 'Meeting:index#establishMediaConnection -->';
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:5812]: Sure, adding ```addMedia():``` to the LOG_HEADERS e.g. const LOG_HEADER = ```'Meeting:index#addMedia():establishMediaConnection -->';```

**Review by marcin-bazyl — COMMENTED** [review]
Great job overall! I've made some only minor comments.

File: docs/samples/webex.min.js
[the remark this answers is not in the document]
Comment (marcin-bazyl) [reply docs/samples/webex.min.js:1]: it's something to do with how SDK is published, the files are autogenerated with each release, for some reason they need to be in git, Arun knows more

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js [hunk 4]
```diff
@@ -2014,481 +2094,540 @@ describe('plugin-meetings', () => {
...
+            meeting.selfUrl = 'selfUrl';
+            meeting.mediaProperties.waitForMediaConnectionConnected = sinon.stub().resolves();
+            meeting.mediaProperties.getCurrentConnectionType = sinon.stub().resolves('udp');
+            meeting.setMercuryListener = sinon.stub();
+            meeting.locusInfo.onFullLocus = sinon.stub();
+            meeting.webex.meetings.geoHintInfo = {regionCode: 'EU', countryCode: 'UK'};
+            meeting.roap.doTurnDiscovery = sinon
+              .stub()
+              .resolves({turnServerInfo: {}, turnDiscoverySkippedReason: 'reachability'});
+            meeting.waitForRemoteSDPAnswer = sinon.stub().resolves();
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js]: I've changed it to simulate the roap OK, and we check that deferSDPAnswer promise is resolved, the timer cleared and ```locusMediaRequestStub``` is called twice, but I couldn't get an assert to work that checked what```locusMediaRequestStub``` was called with the second time (despite console logging what it was being called with i kept getting an assert error)

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 5]
```diff
@@ -5589,28 +5642,285 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    }
+  }
+
+  /**
+   * Enables statsAnalyser if config allows it
+   *
+   * @private
+   * @returns {void}
+   */
+  private enableStatsAnalyzer() {
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: Done

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 6]
```diff
@@ -5589,28 +5642,285 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+        await this.remoteMediaManager.start();
+      }
+
+      await mc.initiateOffer();
+
+      await this.waitForRemoteSDPAnswer();
+
+      this.handleMediaLogging(this.mediaProperties);
+      LoggerProxy.logger.info(`${LOG_HEADER} media connection created`);
```
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/src/meeting/index.ts]: Moved it up

File: packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jor-row) [reply packages/@webex/plugin-meetings/test/unit/spec/meeting/index.js]: I've fixed this now

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 7]
```diff
@@ -5589,28 +5642,290 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+  /**
+   * Returns a promise. This promise is created once the local sdp offer has been successfully created and is resolved
+   * once the remote sdp answer has been received.
+   *
+   * @private
+   * @returns {Promise<void>}
+   */
+  private async waitForRemoteSDPAnswer(): Promise<void> {
+    const LOG_HEADER = 'Meeting:index#addMedia():waitForRemoteSDPAnswer -->';
```
Comment (chrisadubois) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:5777]: I'm not so sure I'm a fan of using the addMedia() as the log header for each of these functions. I think it's still an improvement on whats there now, but it does make it tricky in the future.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 8]
```diff
@@ -5589,28 +5642,290 @@ export default class Meeting extends StatelessWebexPlugin {
...
+
+  /**
+   * Does TURN discovery, SDP offer/answer exhange, establishes ICE connection and DTLS handshake
+   *
+   * @private
+   * @param {RemoteMediaManagerConfiguration} [remoteMediaManagerConfig]
+   * @param {BundlePolicy} [bundlePolicy]
+   * @returns {Promise<void>}
+   */
+  private async establishMediaConnection(
```
Comment (chrisadubois) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:5811]: I know there's a lot of references to `this` in these function calls and it makes it tricky, but part of me feels like all of these functions should be abstracted to like a meeting/media utils file, or a new class that has all of these new split functions defined. There's pros and cons to this, however so it's not a blocker but something we might wanna consider.

File: packages/@webex/plugin-meetings/src/meeting/index.ts [hunk 9]
```diff
@@ -5589,28 +5642,290 @@ export default class Meeting extends StatelessWebexPlugin {
...
+    await this.waitForMediaConnectionConnected();
+  }
+
+  /**
+   * Cleans up stats analyzer, peer connection, and turns off listeners
+   *
+   * @private
+   * @returns {Promise<void>}
+   */
+  private async cleanUpOnAddMediaFailure() {
```
Comment (chrisadubois) [inline packages/@webex/plugin-meetings/src/meeting/index.ts:5894]: v nice

**Review by chrisadubois — APPROVED** [review]
tests look good, code looks great, a couple of food for thought comments for the future for all of us.

File: packages/@webex/plugin-meetings/src/meeting/index.ts (hunk 7 above)
Comment (marcin-bazyl) [reply packages/@webex/plugin-meetings/src/meeting/index.ts:5777]: I think it's important to have a common string in all the functions that get called from addMedia() - this makes it very easy to search the logs for addMedia() related issues and these functions are not called from anywhere else anyway, if any of them does get called from other place then that "addMedia" string can be passed as an argument

**Pull request merged**